### PR TITLE
Kafka source multiple topic with multithreading implementation

### PR DESCRIPTION
**Pull request by ajeeshakd**
### Description

- Created Kafka Source Plugin
- Implementation of multiple-topic with single broker
- Implementation the Json and Plaintext deserializer
- Multithreaded Kafka consumer

 
### Issues Resolved

- Github issue #254

 
### Check List
- [ ] New functionality Kafka Source Plugin.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/MultithreadedConsumer.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		this.plainTextConsumer = new KafkaConsumer<>(consumerProperties);
+
+	}
+
+	@SuppressWarnings({ "unchecked" })
+	@Override
+	public void run() {
+		LOG.info("Start reading ConsumerId:{}, Consumer group: {}", consumerId, consumerGroupId);
+		try {
+			if ((MessageFormat.getByMessageFormatByName(topicConfig.getSchemaConfig().getSchemaType())
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/MultithreadedConsumer.java:59]: Using Switch statement is much better option considering other consumers types will get added and we should also have a default logic if Schema_type is not given in YAML

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/JsonConsumer.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  @SuppressWarnings("unchecked")
+  public void writeToBuffer(final JsonNode jsonNode, Buffer<Record<Object>> buffer) throws TimeoutException {
+    try {
+      final JsonParser jsonParser = jsonFactory.createParser(jsonNode.toString());
+      final Map<String, Object> innerJson = objectMapper.readValue(jsonParser, Map.class);
+      Event event = JacksonLog.builder().withData(innerJson).build();
+      Record<Object> jsonRecord = new Record<>(event);
+      buffer.write(jsonRecord, 1200);
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/JsonConsumer.java:87]: buffer.writeAll can be used here. Please refer SQS/S3 code

Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/JsonConsumer.java:87]: Define 1200 as a constant variable and use it. Can this be configurable in YAML and have a default value ?

File: data-prepper-plugins/kafka-source/README.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+              fetch_max_wait: 500  #optional and dafault is 500
+              fetch_min_bytes: 1  #optional and dafault is 1
+              retry_backoff: 100s  #optional and dafault is 10s
+              max_poll_interval: 300000s  #optional and dafault is 300000s
+              consumer_max_poll_records: 500  #optional and dafault is 500
+            schema:
+              registry_url: http://localhost:8081/
+              key_deserializer: org.apache.kafka.common.serialization.StringDeserializer
+              value_deserializer: org.apache.kafka.common.serialization.StringDeserializer
+              schema_type: plaintext
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/README.md:47]: schema_type: plaintext
 record_type: plaintext

If these are optional and user chooses to not give them, default logic should be plaintext

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          }
+        }
+      }
+    } catch (Exception exp) {
+      LOG.error("Error while reading plain text records from the topic...{}", exp.getMessage());
+    }
+  }
+
+  @Override
+  public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java:73]: onPartitionsAssigned, onPartitionsRevoked and buffer writes can be moved to a common class and used in all the consumer class objects

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSourceBuffer.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * * A helper utility class which helps to write different formats of records
+ * like json and string to the buffer.
+ */
+@SuppressWarnings("deprecation")
+public class KafkaSourceBuffer<K, V> {
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSourceBuffer.java:33]: This call is more of a Util class and not actually a buffer class, can this be renamed appropriately ?

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100));
+        if (!records.isEmpty() && records.count() > 0) {
+          for (TopicPartition partition : records.partitions()) {
+            List<ConsumerRecord<String, String>> partitionRecords = records.records(partition);
+            for (ConsumerRecord<String, String> consumerRecord : partitionRecords) {
+              currentOffsets.put(new TopicPartition(consumerRecord.topic(), consumerRecord.partition()),
+                 new OffsetAndMetadata(consumerRecord.offset() + 1, null));
+              kafkaSourceBuffer.publishRecordToBuffer(consumerRecord.value(), buffer, topicConfig);
+              lastReadOffset = partitionRecords.get(partitionRecords.size() - 1).offset();
+              kafkaSourceBuffer.commitOffsets(partition, lastReadOffset, consumer);
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java:62]: Offsets commits on every read is expensive.  I would have a periodic commits of offsets to improve efficiency. 

Define commitOffsets() method and call it periodically.  Refer good article :  https://www.confluent.io/blog/kafka-consumer-multi-threaded-messaging/

File: data-prepper-plugins/kafka-source/README.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+## Configuration Options
+
+```
+log-pipeline:
+  source:
+    kafka:
+      bootstrap_servers:
+        - 127.0.0.1:9092,localhost:9093, 127.0.0.1:9094
+      topics:
```
Comment (ashoktelukuntla) [inline data-prepper-plugins/kafka-source/README.md:19]: Incorporate latest YAML review comments as discussed for topics and others

**Review by ashoktelukuntla — COMMENTED** [review]
Initial code review comments

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/JsonConsumer.java (hunk 2 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/JsonConsumer.java]: Resolved.
We implemented buffer.writeAll() and moved the writeAll() logic to KafkaSourceBufferAccumulator.java

File: data-prepper-plugins/kafka-source/README.md (hunk 3 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/README.md:47]: Resolved.
by default, the schema_type is PLAINTEXT

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java (hunk 4 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java:82]: Buffer writes for JsonConsumer and PlaintextConsumer has been moved to KafkaSourceBufferAccumulator.java.
In progressing with moving onPartitionsAssigned() and onPartitionsRevoked() to a common class.

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSourceBuffer.java (hunk 5 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSourceBuffer.java]: Resolved.
This class is renamed as KafkaSourceBufferAccumulator.java

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java (hunk 6 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/PlainTextConsumer.java]: Resolved.
As of now we have a periodic offset commits every 5 minutes.

File: data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/MultithreadedConsumer.java (hunk 1 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/consumer/MultithreadedConsumer.java]: Resolved.
Replaced the IF logic with Switch statement and implemented the PLAINTEXT default logic if schema_type is not given in the YAML.

File: data-prepper-plugins/kafka-source/README.md (hunk 7 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-source/README.md:19]: Except the topic related changes, all other changes have been resolved.

We removed the config value 'name' under the topics and assigned the topic-name directly to the -topic variable as shown below.
     ```
      topics:
        - topic: my-topic-2
          auth_type: SASL_OAUTH
          consumer:
              group_name: kafka-consumer-group-2
              group_id: DPKafkaProj-2
```
then we get the below exception:
org.opensearch.dataprepper.parser.ParseException: Failed to parse the configuration file D:\Projects\data-prepper-new\pipelines

**Review by kkondaka — COMMENTED** [review]
I have a high level comment on this. We need the "consumer" code in a directory outside kafka-source because we may be using this code in other places (like kafka buffer) in the future.

**Review by kkondaka — COMMENTED** [review]
Please make a directory `data-prepper-plugins/kafka-plugins` and in side that in the end create separate directories for `configuration`, `consumer` etc

**Comment by ajeeshakd** [comment]
We are closing this PR because we have decided to raise incremental PRs with the same functionality available here.

**Pull request closed without merging**